### PR TITLE
Tforkner/3499 private contractor costs validation

### LINKED ISSUE #3499: [Feature] Update Private Contractor Costs
**Description and related issues -**
Remove the red warning when there is no Private Contractor Cost added.

Update the expanded Private Contractor Costs form. Add a Cancel button. Change the Done button to Add.
no to hourly

yes to hourly

Add in-line validation to Private Contractor Costs form. The validation should appear after the user hits the Add (previously Done) button. If there are validation errors, the errors should display, nothing should be saved, and it should stay on the form until the errors are fixed.

Update the view of added Private Contractor Costs

### This task is done when…
- [ ] remove red warning
- [ ] change Done button to Add
- [ ] add Cancel button to expanded form
- [ ] add in-line validation
- [ ] update to match Figma

### PR DESCRIPTION
**Pull request by thetif**
Resolves #3499 

**Description-**
Added validation to the private contractor form.

**This pull request changes...**

- private contractor form will not save unless the form is valid

**This pull request also touches…**

- adds in testing-library linting

### This pull request is ready to review when...

- [x] Automated tests are updated (and all tests are passing)
- [x] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented

### This pull request can be merged when…

- [ ] Code has been reviewed by someone other than the original author
- [x] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

**Review by tbolt — APPROVED** [review]
Looks great. Nice cleanup to the tests with testing-library

**Comment by itsonlydio** [comment]
| Given  | When | Then| Result |
| ------------- | ------------- | ------------- | ------------- |
| Date field Month/Day | When I entered an invalid date and then a valid date  | the new VALID date should be accepted and save the form  | failed |

This is good the only thing we found is the 'Provide Start Date' warning still appears. 
image

**Comment by mirano-darren** [comment]
Also the contractor view in the dashboard doesn't match Figma for me? Here's what I'm seeing
image

The Figma has a line separating the two
image

**Comment by thetif** [comment]
@itsonlydio and @mirano-darren I think I addressed the issues can you check it again?

**Comment by mirano-darren** [comment]
Update on some of the test cases found:
| Given  | When | Then | Result |
| ------------- | ------------- | ------------- | ------------- |
| A date field  | An invalid month is entered then corrected to a valid month | The newly entered month should be considered valid, with no error message appearing  |  PASS ✔️, but then you cannot save afterwards! FAIL ❌  |
| Private contractor dashboard view  | Multiple contractors have been entered  | A line is separating the contractors | PASS ✔️ |
| A date field  | An invalid day is entered then corrected to a valid day | The newly entered day  should be considered valid, with no error message appearing  |  PASS ✔️, but you cannot save afterwards! FAIL ❌  |
| A date field  | a number greater than 12 is entered for the month  | Save button isn't clickable, | PASS ✔️ |
| A date field  | a number greater than 32 is entered for the day | Save button isn't clickable | PASS ✔️ |
| Both date fields | The end date comes before the start date  | Save button is clickable, no error message appears | FAIL ❌ |
| A date field | When you enter a date for the first time | Whether or not the date is valid, the inline validation error message appears. Retriggering the validation will then accept a valid date | FAIL ❌

**Comment by thetif** [comment]
@mirano-darren can you test this again when you get the chance? I think I got it all fixed this time.

**Review by jeromeleecms — CHANGES_REQUESTED** [review]
The date validation on this field doesn't seem be working (this is seen as valid):

image

Also, saving the form truncates the values to the proper MM-DD-YYYY. I don't know if it's related or not, but the preview site crashed right after...

Number of hours - this should allow 0 hours (consistent with the language in the validation). In some cases, a state may have 0 cost for the first year, but have a cost the 2nd year (or vice versa)

image

FFY Cost should also allow 0 for the same rationale. In this case, the validation should read "Provide an annual cost greater than or equal to 0"

image

**Comment by thetif** [comment]
@jeromeleecms I'm still working allowing 0 on those number, but the problem with the date is that the design system date field truncates the numbers so even though you are entering 111 in month, it's only returning 11 to us, so that looks correct.

**Comment by jeromeleecms** [comment]
Can/should we add a "maxlength" tag to the date fields? I vaguely remember they introduced that in HTML5 (seems like there is a react equivalent of "maxLength") or does the design system not want us using that or does that cause other problems?

**Comment by thetif** [comment]
we don't have access to those fields, the best we could do is put in a ticket with design system

**Comment by thetif** [comment]
also, we previously didn't allow 0s in any number fields, but I've opened it up so that we can use 0

**Review by beparticular — APPROVED** [review]

**Review by tbolt — APPROVED** [review]

File: web/src/components/DateField.js [hunk 1]
```diff
@@ -1,104 +1,177 @@
...
+            [action.field]: action.value
+          }
+        };
+      case 'yearValidate': {
+        const { value: year } = action;
+        return {
+          ...state,
+          invalidObject: {
+            ...state.invalidObject,
+            yearInvalid: !isNumeric(year) || +year < 1900 || +year > 2100
```
Comment (tbolt) [inline web/src/components/DateField.js:78]: this is nice!

**Review by tbolt — APPROVED** [review]
looks good. nice work

**Review by itsonlydio — COMMENTED** [review]
As I was testing the in-line validation works in the date fields, Contract costs, hourly resource fields and FFY Cost FIeld.
Screen Shot 2022-05-02 at 12 24 46 PM
Screen Shot 2022-05-02 at 12 24 54 PM
Screen Shot 2022-05-02 at 12 25 06 PM

**Review by jeromeleecms — APPROVED** [review]
Looks good!

**Pull request merged**